### PR TITLE
Refine SDP pipeline

### PR DESCRIPTION
**Pull request by live1206**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4855)

### CODE REVIEW

File: src/docfx/build/page/BuildPage.cs [hunk 1]
```diff
@@ -16,33 +16,24 @@ public static async Task<List<Error>> Build(Context context, Document file)
         {
             Debug.Assert(file.ContentType == ContentType.Page);
 
-            var (errors, pageModel, inputMetadata) = await Load(context, file);
-            var (generateErrors, outputMetadata) = await GenerateOutputMetadata(context, file, inputMetadata);
-            errors.AddRange(generateErrors);
+            var (errors, pageModel) = await Load(context, file);
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: Need to clarify the terms used here, shall we call the original JSON `sourceModel`, input to template `pageModel`, and the current model (after JINT/mustache) stored in DHS `templateModel` ?

```suggestion
            var (errors, sourceModel) = await Load(context, file);
```

File: src/docfx/build/page/BuildPage.cs [hunk 2]
```diff
@@ -16,33 +16,24 @@ public static async Task<List<Error>> Build(Context context, Document file)
...
+            var (monikerError, monikers) = context.MonikerProvider.GetFileLevelMonikers(file);
+            errors.AddIfNotNull(monikerError);
 
-            var outputPath = file.GetOutputPath(outputMetadata.Monikers, file.Docset.SiteBasePath, file.IsPage);
+            var outputPath = file.GetOutputPath(monikers, file.Docset.SiteBasePath, file.IsPage);
 
             object output = null;
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: consider a function over if else:
```suggestion
            var (outputErrors, output, metadata) = file.IsPage ? await CreatePageOutput() : CreateDataOutput();
```

File: src/docfx/build/page/BuildPage.cs [hunk 3]
```diff
@@ -83,7 +74,39 @@ public static async Task<List<Error>> Build(Context context, Document file)
             return errors;
         }
 
-        private static async Task<(List<Error>, OutputMetadata)> GenerateOutputMetadata(
+        private static async Task<(List<Error> errors, object output, JObject metadata)> CreatePageOutput(Context context, Document file, JObject model)
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: consider break it, this line is too long. I am looking for a tool to enforce max line with to be something like 140
```suggestion
        private static async Task<(List<Error> errors, object output, JObject metadata)>
            CreatePageOutput(Context context, Document file, JObject model)
```

File: src/docfx/build/page/BuildPage.cs [hunk 4]
```diff
@@ -220,75 +240,58 @@ private static (List<Error> errors, JObject model, InputMetadata inputMetadata)
...
+                    ExtensionData = landingData.ExtensionData,
                 });
             }
 
-            return (errors, pageModel, inputMetadata);
+            return (errors, pageModel);
         }
 
-        private static (object model, JObject metadata) ApplyPageTemplate(Context context, Document file, JObject pageMetadata, JObject pageModel, bool isConceptual)
+        private static (TemplateModel model, JObject metadata) TransformToTemplateModel(Context context, JObject pageModel, Document file)
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]: consider `CreateXXX` to align with other `CreateXXX` methods

```suggestion
        private static (TemplateModel model, JObject metadata) CreateTemplateModel(Context context, JObject pageModel, Document file)
```

File: src/docfx/build/page/BuildPage.cs (hunk 1 above)
Comment (live1206) [reply src/docfx/build/page/BuildPage.cs]: sounds good to me

File: src/docfx/build/page/BuildPage.cs [hunk 5]
```diff
@@ -83,7 +64,47 @@ public static async Task<List<Error>> Build(Context context, Document file)
             return errors;
         }
 
-        private static async Task<(List<Error>, OutputMetadata)> GenerateOutputMetadata(
+        private static async Task<(List<Error> errors, object output, JObject metadata)>
+            CreatePageOutput(Context context, Document file, JObject sourcecModel)
```
Comment (yufeih) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
            CreatePageOutput(Context context, Document file, JObject sourceModel)
```

File: src/docfx/template/MustacheTemplate.cs [hunk 6]
```diff
@@ -37,11 +37,19 @@ RendererSettingsBuilder UseJson(RendererSettingsBuilder settings)
             }
         }
 
-        public string Render(string templateName, JToken model)
+        public string Render(string templateFileName, JToken model)
```
Comment (yufeih) [inline src/docfx/template/MustacheTemplate.cs]: the mustache template wasn't run, based on diff, can you add a spec test for mustache template?

File: src/docfx/build/page/BuildPage.cs [hunk 7]
```diff
@@ -83,7 +64,47 @@ public static async Task<List<Error>> Build(Context context, Document file)
...
+            var pageModel = new JObject();
+            JsonUtility.Merge(pageModel, sourcecModel);
+            JsonUtility.Merge(pageModel, mergedMetadata);
+
+            if (file.Docset.Config.Output.Json && !file.Docset.Legacy)
+            {
+                return (errors, pageModel, SortProperties(mergedMetadata));
+            }
+
+            var (templateModel, metadata) = CreateTemplateModel(context, new JObject(pageModel.Properties().OrderBy(p => p.Name)), file);
```
Comment (OsmondJiang) [inline src/docfx/build/page/BuildPage.cs]:
```suggestion
var (templateModel, metadata) = CreateTemplateModel(context, SortProperties(pageModel), file);
```

File: src/docfx/template/TemplateEngine.cs [hunk 8]
```diff
@@ -54,29 +57,25 @@ public TemplateSchema GetSchema(SourceInfo<string> schemaName)
...
             var themeRelativePath = PathUtility.GetRelativePathToFile(file.SitePath, "_themes");
 
             var liquidModel = new JObject
             {
-                ["content"] = content,
-                ["page"] = rawMetadata,
-                ["metadata"] = metadata,
+                ["content"] = model.Content,
+                ["page"] = model.RawMetadata,
+                ["metadata"] = new JObject(),
```
Comment (OsmondJiang) [inline src/docfx/template/TemplateEngine.cs]: why it's empty JObject?

File: src/docfx/template/TemplateEngine.cs [hunk 9]
```diff
@@ -98,13 +97,22 @@ public void CopyTo(string outputPath)
...
         {
-            var scriptPath = Path.Combine(_templateDir, "ContentTemplate", scriptName);
+            var scriptPath = Path.Combine(_contentTemplateDir, scriptName);
             if (!File.Exists(scriptPath))
             {
                 return model;
             }
 
-            var result = (JObject)_js.Run(scriptPath, methodName, model);
+            var jsResult = _js.Run(scriptPath, methodName, model);
```
Comment (yufeih) [inline src/docfx/template/TemplateEngine.cs]: I'm patching docs-ui here https://dev.azure.com/ceapex/Engineering/_git/docs-ui/pullrequest/13489?_a=overview, could be removed after the patch is deployed.

File: src/docfx/template/TemplateEngine.cs (hunk 8 above)
Comment (live1206) [reply src/docfx/template/TemplateEngine.cs]: because it is not needed, @yufeih can we just remove `["metadata"]`?

Comment (yufeih) [reply src/docfx/template/TemplateEngine.cs]: @live1206 , checked it twice, it should be:
```suggestion
                ["metadata"] = model.PageMetadata,
```

File: src/docfx/template/MustacheTemplate.cs (hunk 6 above)
Comment (live1206) [reply src/docfx/template/MustacheTemplate.cs]: added a spec test to verify SDP pipeline

**Comment by live1206** [comment]
> nit: describe a little bit about what the pipeline does

Yep, I had a thought about it. Feel like we should have a spec doc for it?

**Pull request merged**